src/sync/sync_permission.hpp

```diff
@@ -0,0 +1,149 @@
+// APIs. They are immutable objects.
+struct Permission {
+    // The path of the Realm to which this permission pertains.
+    std::string path;
+
+    // A permission encapsulates a single access level.
+    // Each level includes all the capabilities of the level
+    // above it (for example, 'write' implies 'read').
+    enum class AccessLevel {
+        None,
+        Read,
+        Write,
+        Admin,
+    };
+    AccessLevel access;
+
+    // Condition is a userId or a KeyValue pair
+    // Other conditions may be supported in the future
+    struct Condition {
+        enum class Type {
+            // The permission is applied to a single user based on their user ID
+            UserId,
+            // The permission is based on any user that meets a criterion specified by key/value.
+            KeyValue,
+        };
+        Type type;
+
+        // FIXME: turn this back into a union type
+        std::string user_id = "";
+        std::pair<std::string, std::string> key_value = std::make_pair("", "");
+
+        Condition(std::string id)
+        : type(Type::UserId)
+        , user_id(std::move(id))
+        { }
+
+        Condition(std::string key, std::string value)
+        : type(Type::KeyValue)
+        , key_value(std::make_pair(std::move(key), std::move(value)))
+        { }
+    };
+    Condition condition;
+};
+
+class PermissionResults {
+public:
+    // The number of permissions represented by this PermissionResults.
+    size_t size()
+    {
+        return m_results.size();
+    }
+
+    // Get the permission value at the given index.
+    // Throws an `OutOfBoundsIndexException` if the index is invalid.
+    Permission get(size_t index);
+
+    // Create an async query from this Results.
+    // The query will be run on a background thread and delivered to the callback,
+    // and then rerun after each commit (if needed) and redelivered if it changed
+    NotificationToken async(std::function<void(std::exception_ptr)> target)
+    {
+        return m_results.async(std::move(target));
+    }
+
+    // Create a new instance by further filtering or sorting this instance.
+    PermissionResults filter(Query&& q) const
+    {
+        return PermissionResults(m_results.filter(std::move(q)));
+    }
+
+    // Don't use this constructor directly. Publicly exposed so `make_unique` can see it.
+    PermissionResults(Results&& results)
+    : m_results(results)
+    { }
+
+protected:
+    Results m_results;
+};
+
+class Permissions {
+public:
+    // Consumers of these APIs need to pass in a method which creates a Config with the proper
+    // SyncConfig and associated callbacks, as well as the path and other parameters.
+    using ConfigMaker = std::function<Realm::Config(std::shared_ptr<SyncUser>, std::string url)>;
+
+    // Callback used to asynchronously vend a `PermissionResults` object.
+    using PermissionResultsCallback = std::function<void(std::unique_ptr<PermissionResults>, std::exception_ptr)>;
```

Comment:
Why does this take a unique pointer to a `PermissionResults`? If it's optional because there may have been an error then it should use `util::Optional` instead.